modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java

```diff
@@ -6295,6 +6310,33 @@ void addMessage(TcpDiscoveryAbstractMessage msg, @Nullable byte[] msgBytes) {
                 log.debug("Message has been added to client queue: " + msg);
         }
 
+
+        /**
+         * Check the last time a heartbeat message received.
+         * In case of timeout, expel client node from the topology
+         */
+        @Override
+        protected void noMessageLoop() {
+            long period = U.currentTimeMillis() - lastHeartBeatTime;
+
+            if (period >= heartBeatTimeOut) {
+                if (log.isInfoEnabled())
+                    log.info("### No heartbeat message from node:" + clientNodeId + "; timeOut=" + heartBeatTimeOut + "; period=" + period);
```

Comment:
We have to use another pattern:
throw new CacheException("Invalid type [type=" + type + ", xid=" + tx.getXid() + ']');
throw new WorkflowException("Invalid rule: " + rule.getName());
Check https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines for details